homeassistant/components/weather/__init__.py

```diff
@@ -115,36 +190,84 @@ class WeatherEntity(Entity):
     _attr_ozone: float | None = None
     _attr_precision: float
     _attr_pressure: float | None = None
-    _attr_pressure_unit: str | None = None
+    _attr_pressure_unit: None = None  # Subclasses of WeatherEntity should not set this
+    _attr_native_pressure_unit: str | None = None
     _attr_state: None = None
-    _attr_temperature_unit: str
+    _attr_temperature_unit: None = (
+        None  # Subclasses of WeatherEntity should not set this
+    )
+    _attr_native_temperature_unit: str
     _attr_temperature: float | None
     _attr_visibility: float | None = None
-    _attr_visibility_unit: str | None = None
-    _attr_precipitation_unit: str | None = None
+    _attr_visibility_unit: None = (
+        None  # Subclasses of WeatherEntity should not set this
+    )
+    _attr_native_visibility_unit: str | None = None
+    _attr_precipitation_unit: None = (
+        None  # Subclasses of WeatherEntity should not set this
+    )
+    _attr_native_precipitation_unit: str | None = None
     _attr_wind_bearing: float | str | None = None
     _attr_wind_speed: float | None = None
-    _attr_wind_speed_unit: str | None = None
+    _attr_wind_speed_unit: None = (
+        None  # Subclasses of WeatherEntity should not set this
+    )
+    _attr_native_wind_speed_unit: str | None = None
+
+    _weather_option_temperature_uom: str | None = None
+    _weather_option_pressure_uom: str | None = None
+    _weather_option_visibility_uom: str | None = None
+    _weather_option_precipitation_uom: str | None = None
+    _weather_option_wind_speed_uom: str | None = None
+
+    async def async_internal_added_to_hass(self) -> None:
+        """Call when the sensor entity is added to hass."""
+        await super().async_internal_added_to_hass()
+        if not self.registry_entry:
+            return
+        self.async_registry_entry_updated()
 
     @property
     def temperature(self) -> float | None:
         """Return the platform temperature in native units (i.e. not converted)."""
         return self._attr_temperature
 
     @property
-    def temperature_unit(self) -> str:
+    def native_temperature_unit(self) -> str | None:
         """Return the native unit of measurement for temperature."""
-        return self._attr_temperature_unit
+        if hasattr(self, "_attr_native_temperature_unit"):
+            return self._attr_native_temperature_unit
+        return None
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the unit of measurement for temperature."""
+        if self._weather_option_temperature_uom:
+            return self._weather_option_temperature_uom
+
+        return self.hass.config.units.temperature_unit
```

Comment:
This is OK, weather entity already converted temperatures, but we should probably return `None` here if `self.native_temperature_unit` is `None`.